### PR TITLE
Hw2 task03

### PR DESCRIPTION
**Pull request by mirmozavr**

### CODE REVIEW

File: hw2/task03/hw2_task03.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    [1, 4],
+    [2, 3],
+    [2, 4],
+]
+"""
+import itertools
+
+
+def combinations(*args: list) -> list:
+    return list(itertools.product(*args))
```
Comment (Cjkjvfnby) [inline hw2/task03/hw2_task03.py]: Please solve it without `itertools` module.

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

File: hw2/task03/hw2_task03.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    [1, 3],
+    [1, 4],
+    [2, 3],
+    [2, 4],
+]
+"""
+
+
+def combination_of_lists(*args: list) -> list:
+    pools = [tuple(pool) for pool in args]
```
Comment (Cjkjvfnby) [inline hw2/task03/hw2_task03.py]: Why do you convert args to pools?

File: hw2/task03/hw2_task03.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    result = [[]]
+    for pool in pools:
+        result = [x + [y] for x in result for y in pool]
```
Comment (Cjkjvfnby) [inline hw2/task03/hw2_task03.py]: Please give x and y meaningful names.

File: hw2/task03/test_hw2_task03.py [hunk 4]
```diff
@@ -0,0 +1,8 @@
+from hw2_task03 import combination_of_lists
+
+
+def test_combination_of_lists():
+    assert combination_of_lists([1, 2], [3, 4]) == [[1, 3], [1, 4], [2, 3], [2, 4]]
```
Comment (Cjkjvfnby) [inline hw2/task03/test_hw2_task03.py:5]: You K is always 2.

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

File: hw2/task03/hw2_task03.py (hunk 2 above)
Comment (mirmozavr) [reply hw2/task03/hw2_task03.py]: No need for that with lists as args

File: hw2/task03/hw2_task03.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    [2, 3],
+    [2, 4],
+]
+"""
+
+
+def combination_of_lists(*args: list) -> list:
+    result = [[]]
+    for pool in args:
+        result = [base + [suffix] for base in result for suffix in pool]
```
Comment (Cjkjvfnby) [inline hw2/task03/hw2_task03.py:20]: Optional question: how this approach is called in general?

**Review by Cjkjvfnby — APPROVED** [review]

File: hw2/task03/hw2_task03.py (hunk 5 above)
Comment (mirmozavr) [reply hw2/task03/hw2_task03.py:20]: List comprehension and nested loops?

Comment (Cjkjvfnby) [reply hw2/task03/hw2_task03.py:20]: I mean more general term, like `recursion`.

**Pull request merged**